lib/rules/block-attributes-order.js

```diff
@@ -0,0 +1,320 @@
+/**
+ * @fileoverview enforce ordering of block attributes
+ * @author Wenlu Wang
+ */
+'use strict'
+const utils = require('../utils')
+const { parseSelector } = require('../utils/selector')
+
+/**
+ * @enum {string}
+ */
+const WELL_KNOWN_TEMPLATE_ATTRS = {
+  functional: 'functional',
+  lang: 'lang',
+  src: 'src'
+}
+
+/**
+ * @enum {string}
+ */
+const WELL_KNOWN_SCRIPT_ATTRS = {
+  lang: 'lang',
+  setup: 'setup',
+  src: 'src'
+}
+
+/**
+ * @enum {string}
+ */
+const WELL_KNOWN_STYLE_ATTRS = {
+  scoped: 'scoped',
+  module: 'module',
+  lang: 'lang',
+  src: 'src'
+}
+
+/**
+ * @typedef UserOrderOption
+ * @property { string } element
+ * @property { (string | string[])[] } order
+ */
+
+/**
+ * @typedef {import('../utils/selector').VElementSelector} VElementSelector
+ */
+
+/**
+ * @typedef OrderOption
+ * @property {string} element
+ * @property {(string | string[])[]} order
+ * @property {(RegExp | RegExp[])[]} regexOrder
+ * @property {VElementSelector} selector
+ */
+
+/**
+ * @type {readonly UserOrderOption[]}
+ */
+const defaultOptions = Object.freeze([
+  {
+    element: 'style',
+    order: [
+      WELL_KNOWN_STYLE_ATTRS.module,
+      WELL_KNOWN_STYLE_ATTRS.scoped,
+      WELL_KNOWN_STYLE_ATTRS.lang,
+      WELL_KNOWN_STYLE_ATTRS.src
+    ]
+  },
+  {
+    element: 'script',
+    order: [
+      WELL_KNOWN_SCRIPT_ATTRS.setup,
+      WELL_KNOWN_SCRIPT_ATTRS.lang,
+      WELL_KNOWN_SCRIPT_ATTRS.src
+    ]
+  },
+  {
+    element: 'template',
+    order: [
+      WELL_KNOWN_TEMPLATE_ATTRS.functional,
+      WELL_KNOWN_TEMPLATE_ATTRS.lang,
+      WELL_KNOWN_TEMPLATE_ATTRS.src
```

Comment:
After changing it to accept complex ordering, could you change it so that `lang` and `src` are always ordered last?

e.g.

```suggestion
      "functional",
      ["!lang", "!src"],
      "lang",
      "src"
```

Change other elements in the same way.